CRM/Contact/Selector.php

```diff
@@ -927,8 +927,22 @@ public function addActions(&$rows) {
     $mapMask = $mask & 4095;
 
     $links = self::links($this->_context, $this->_contextMenu, $this->_key);
+    $editlink = $links[CRM_Core_Action::UPDATE];  // Remember the Edit link
 
     foreach ($rows as $id => & $row) {
+      //
+      // CRM-12645 Check if we have permission to edit the record.  If we have, we enable the link, otherwise
+      // we don't.
+      if (CRM_Contact_BAO_Contact_Permission::allow($row['contact_id'], CRM_Core_Permission::EDIT)) {
```

Comment:
> This function checks the logged in user for edit all contacts permission and returns TRUE irrespective of the contact_id passed

As @spalmstr  pointed out, I don't think that's quite right, have a look at  https://github.com/civicrm/civicrm-core/blob/4.7.12/CRM/Contact/BAO/Contact/Permission.php#L35-L79